bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ManagedThingProvider.java

```diff
@@ -41,6 +75,159 @@ protected String keyToString(ThingUID key) {
+
+        if (aAcceptedItemType != null) {
+            if (!aAcceptedItemType.equals(bAcceptedItemType)) {
+                return false;
+            }
+        } else if (bAcceptedItemType != null) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private void createThingFromStorageForThingHandlerFactory(Thing storedThing, ThingHandlerFactory factory) {
+        logger.debug("createThingFromStorageForThingHandlerFactory({}, {}", storedThing.getUID(),
+                factory.getClass().getName());
+        if (!loadedXmlThingTypes.contains(getBundleName(factory))) {
+            return;
+        }
+
+        logger.info("*** thing.getHandler() = {}", storedThing.getHandler());
+
+        if (factory.supportsThingType(storedThing.getThingTypeUID())) {
+            Configuration storedConfiguration = storedThing.getConfiguration();
+
+            ThingBuilder thingBuilder = ThingBuilder.create(storedThing.getThingTypeUID(), storedThing.getUID())
+                    .withBridge(storedThing.getBridgeUID()).withChannels(storedThing.getChannels())
+                    .withConfiguration(storedConfiguration).withLabel(storedThing.getLabel())
+                    .withLocation(storedThing.getLocation()).withProperties(storedThing.getProperties());
+
+            Thing factoryThing = factory.createThing(storedThing.getThingTypeUID(), storedConfiguration,
+                    storedThing.getUID(), storedThing.getBridgeUID());
+
+            if (factoryThing != null) {
+                Map<ChannelUID, Channel> factoryChannels = factoryThing.getChannels().stream()
+                        .collect(Collectors.toMap(Channel::getUID, channel -> channel));
+                Map<ChannelUID, Channel> storedChannels = storedThing.getChannels().stream()
+                        .collect(Collectors.toMap(Channel::getUID, channel -> channel));
+
+                // check if we have new or updated channels
+                factoryChannels.forEach((factoryChannelUID, factoryChannel) -> {
+                    if (!storedChannels.containsKey(factoryChannelUID)) {
+                        thingBuilder.withChannel(factoryChannel);
+                        logger.trace("added channel {} to thing {}", factoryChannel, storedThing.getUID());
+                    } else {
+                        if (!compareChannels(factoryChannel, storedThing.getChannel(factoryChannelUID.getId()))) {
+                            thingBuilder.withoutChannel(factoryChannelUID).withChannel(factoryChannel);
+                            logger.trace("updated channel {} in thing {}", factoryChannel, storedThing.getUID());
+                        }
+                    }
+                });
+
+                // check if we have removed channels
+                storedChannels.forEach((storedChannelUID, storedChannel) -> {
+                    if (!factoryChannels.containsKey(storedChannelUID)) {
+                        thingBuilder.withoutChannel(storedChannelUID);
```

Comment:
The reason why I didn’t implement that line is: it will remove user-added channels. As far as I know there is no way to find out if an additional channel was added by the user as `extensible` or if it was present in the XML before and was removed now.

Reply:
I didn't think about user added channels to be honest. Maybe we could mark the channels somehow whether they are user created?

Reply:
That's exactly what I was looking for in my initial comment in [#6537](https://github.com/eclipse/smarthome/pull/6537) but I didn't find it. 

Maybe `Channel` could get a `private boolean userAddedExtensible=false;` with additional setter/getter. This could be set to true if the user adds a channel via UI, serialized for storage and read back from the database later.

Reply:
You have the thing description that tells you about extensible channels. Just filter them out.

Reply:
We would need to add the ThingTypeRegistry then. But that is also fine for the `GenericThingProvider`.